File: day03/src/main.py
import re
from pathlib import Path


def part1():
    memory = Path('my_input.txt').read_text()
    pattern = r'mul\((\d{1,3}),(\d{1,3})\)'

    total_sum = sum(
        int(match.group(1)) * int(match.group(2))
        for match
        in re.finditer(pattern, memory)
    )

    print(f'Answer: {total_sum}')


def part2():
    memory = Path('my_input.txt').read_text()
    pattern = r'mul\((\d{1,3}),(\d{1,3})\)|(do\(\)|don\'t\(\))'
    mul_enabled = True
    total_sum = 0

    for match in re.finditer(pattern, memory):
        if match.group(1) and match.group(2) and mul_enabled:
            total_sum += int(match.group(1)) * int(match.group(2))
        elif match.group(3):
            mul_enabled = match.group(3) == 'do()'

    print(f'Answer: {total_sum}')


if __name__ == '__main__':
    part1()
    part2()

File: day03/src/main.rs
use regex::Regex;
use std::fs;

fn part1() {
    let memory = fs::read_to_string("my_input.txt").expect("Failed to read file");
    let pattern = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)").unwrap();
    let total_sum: i32 = pattern
        .captures_iter(memory.trim())
        .map(|cap| {
            let x: i32 = cap[1].parse().unwrap();
            let y: i32 = cap[2].parse().unwrap();
            x * y
        })
        .sum();
    println!("Answer: {}", total_sum);
}

fn part2() {
    let memory = fs::read_to_string("my_input.txt").expect("Failed to read file");
    let pattern = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)|(do\(\)|don't\(\))").unwrap();
    let mut mul_enabled = true;
    let mut total_sum = 0;

    for cap in pattern.captures_iter(memory.trim()) {
        if mul_enabled {
            if let (Some(x), Some(y)) = (cap.get(1), cap.get(2)) {
                let x: i32 = x.as_str().parse().unwrap();
                let y: i32 = y.as_str().parse().unwrap();
                total_sum += x * y;
            }
        }
        if let Some(control) = cap.get(3) {
            mul_enabled = control.as_str() == "do()";
        }
    }

    println!("Answer: {}", total_sum);
}

fn main() {
    part1();
    part2();
}
